File: src/main/scala/com/kirkwoodwest/scalawoods/utils/SerializationUtil.scala
package com.kirkwoodwest.scalawoods.utils

import java.io._
import java.util
import java.util.Base64


import java.io._
import java.util.Base64
import scala.util.{Try, Using}
import scala.collection.mutable.ListBuffer
import scala.jdk.CollectionConverters._

object SerializationUtil {

  def serialize(o: Serializable): String = {
    Using(new ByteArrayOutputStream()) { baos =>
      Using(new ObjectOutputStream(baos)) { oos =>
        oos.writeObject(o)
        oos.flush()
        Base64.getEncoder.encodeToString(baos.toByteArray)
      }.getOrElse("")
    }.getOrElse("")
  }

  def deserialize(s: String): Option[AnyRef] = {
    val data = Base64.getDecoder.decode(s)
    Using(new ObjectInputStream(new ByteArrayInputStream(data))) { ois =>
      ois.readObject()
    }.toOption
  }

  def deserializeAsList[T](serialized: String, clazz: Class[T]): Option[List[T]] = {
    deserialize(serialized) match {
      case Some(deserializedObject: java.util.List[_]) =>
        Try {
          deserializedObject.asScala.toList.collect {
            case item if clazz.isInstance(item) => clazz.cast(item)
          }
        }.toOption
      case _ =>
        None
    }
  }
}

File: src/main/java/com/kirkwoodwest/openwoods/utils/SerializationUtil.java
package com.kirkwoodwest.openwoods.utils;

import java.io.*;
import java.util.ArrayList;
import java.util.Base64;

public class SerializationUtil {

	public static String serialize(Serializable o) {
		try {
			ByteArrayOutputStream baos = new ByteArrayOutputStream();
			ObjectOutputStream oos = new ObjectOutputStream(baos);

			oos.writeObject(o);
			oos.close();
			return Base64.getEncoder().encodeToString(baos.toByteArray());
		} catch (IOException e) {
			e.printStackTrace();
		}

		return "";
	}

	public static Object deserialize(String s) {
		try {
			byte[] data = Base64.getDecoder().decode(s);
			ObjectInputStream ois = null;
			ois = new ObjectInputStream(new ByteArrayInputStream(data));
			Object o = ois.readObject();
			ois.close();
			return o;
		} catch (IOException | ClassNotFoundException e) {
			e.printStackTrace();
		}
		return null;
	}

	// Generic method to deserialize a serialized string back to an ArrayList of type T
	public static <T> ArrayList<T> deserializeAsList(String serialized, Class<T> clazz) {
		try {
			Object deserializedObject = deserialize(serialized);
			if (deserializedObject instanceof ArrayList<?>) {
				ArrayList<?> arrayList = (ArrayList<?>) deserializedObject;
				ArrayList<T> result = new ArrayList<>();
				for (Object item : arrayList) {
					if (clazz.isInstance(item)) {
						result.add(clazz.cast(item));
					} else {
						throw new ClassCastException("Deserialized list contained an object that is not of type " + clazz.getSimpleName());
					}
				}
				return result;
			} else {
				throw new ClassCastException("Deserialized object is not an ArrayList");
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		return null;
	}
}
